File: src/main/java/com/meti/core/Err.java
package com.meti.core;

import java.util.function.Consumer;
import java.util.function.Function;

/**
 * Represents an error type.
 */
public record Err<T, E extends Throwable>(E inner) implements Result<T, E> {
    public static <T, E extends Throwable> Result<T, E> apply(E inner) {
        return new Err<>(inner);
    }

    @Override
    public Option<T> value() {
        return None.apply();
    }

    @Override
    public Option<E> err() {
        return Some.apply(inner);
    }

    @Override
    public <R> Result<R, E> mapValue(Function<T, R> mapper) {
        return apply(inner);
    }

    @Override
    public <R> Result<R, E> mapValueToResult(Function<T, Result<R, E>> mapper) {
        return apply(inner);
    }

    @Override
    public void consume(Consumer<T> valueConsumer, Consumer<E> errorConsumer) {
        try {
            errorConsumer.accept(inner);
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    @Override
    public <R> R into(Function<Result<T, E>, R> mapper) {
        return mapper.apply(this);
    }

    @Override
    public <R> R match(Function<T, R> onOk, Function<E, R> onErr) {
        return onErr.apply(inner);
    }

    @Override
    public <R extends Throwable> Result<T, R> mapErr(Function<E, R> mapper) {
        return new Err<>(mapper.apply(inner));
    }

    @Override
    public Result<T, E> peekValue(Consumer<T> consumer) {
        return this;
    }

    @Override
    public T $() throws E {
        throw inner;
    }
}

File: src/windows/magma/api/result/Err.c
#include "./Err.h"
export class Err<T, X> implements Result<T, X> {
	X error;
	constructor (X error) {
		this.error = error;
	}
}

Option<X> findError() {
	return new Some<X>(this/*auto*/.error);
}
Option<T> findValue() {
	return new None<T>(/*auto*/);
}
R match((arg0 : T) => R whenOk, (arg0 : X) => R whenErr) {
	return whenErr/*(arg0 : X) => R*/(this/*auto*/.error);
}
Result<R, X> flatMapValue((arg0 : T) => Result<R, X> mapper) {
	return new Err<>(this/*auto*/.error);
}
Result<R, X> mapValue((arg0 : T) => R mapper) {
	return new Err<>(this/*auto*/.error);
}